### PR TITLE
Prometheus Frigate stats exporter #5463

### PR DESCRIPTION
**Pull request by mitchross**
Merge the work from https://github.com/blakeblackshear/frigate/discussions/5463 into Frigate 

Addresses https://github.com/blakeblackshear/frigate/issues/2266

### CODE REVIEW

**Comment by mitchross** [comment]
@NickM-27  unit tests run fine locally for me... any ideas why build is held up?

**Comment by blakeblackshear** [comment]
Google's gpg key has been returning a 500 most of the day. This is probably causing the build to fail upstream of the tests.

File: frigate/http.py
Comment (NickM-27) [inline frigate/http.py:88]: I think there may be a better name than wsgi_app for this since right now it is Prometheus specific?

File: frigate/prometheus_exporter.py
Comment (NickM-27) [inline frigate/prometheus_exporter.py:1]: I think it might be good to go ahead and create a `stats` folder and include the stats.py as well as this file (maybe with different names) in that folder.

File: frigate/http.py
Comment (mitchross) [reply frigate/http.py:88]: Was following instructions here https://github.com/prometheus/client_python#flask

I believe it comes from from prometheus_client import make_wsgi_app

File: frigate/prometheus_exporter.py
Comment (mitchross) [reply frigate/prometheus_exporter.py:1]: I think this makes sense

File: frigate/http.py
Comment (NickM-27) [reply frigate/http.py:88]: Right, but that's an example. I believe the variable name can be whatever we want as the function and variable name shouldn't be dependent.

Comment (mitchross) [reply frigate/http.py:88]: Ah ok, when I get home from work i'll try this. Any name suggestions?

Comment (NickM-27) [reply frigate/http.py:88]: Maybe `app.prometheus_proxy`

Comment (mitchross) [reply frigate/http.py:88]: I cant remove the wsgi_app part because its part of Flask

image


if I do something like what you suggest

image

I get 

image


thoughts?

Comment (NickM-27) [reply frigate/http.py:88]: Interesting, not what I expected. In any case thanks for trying it.

File: frigate/prometheus_exporter.py
Comment (mitchross) [reply frigate/prometheus_exporter.py:1]: Any Name for stats class? We will get 
image

Comment (NickM-27) [reply frigate/prometheus_exporter.py:1]: My personal vote would be to just call it frigate (so it would be `frigate.stats.frigate` but maybe there is a better name

Comment (mitchross) [reply frigate/prometheus_exporter.py:1]: @blakeblackshear  if you have any opinions please share. I'm not a daily python user, so not familiar with conventions.

Comment (mitchross) [reply frigate/prometheus_exporter.py:1]: How about a root folder called monitoring, with two files, "stats.py" and "prometheus_exporter.py" ?

Comment (mweinelt) [reply frigate/prometheus_exporter.py:1]: When applications use the Prometheus libraries themselves, the proper term would be instrumentation, not exporter.

File: frigate/prometheus_exporter.py
Comment (mitchross) [reply frigate/prometheus_exporter.py:33]: Core of the work came from https://github.com/bairhys/prometheus-frigate-exporter , I just integrated it into frigate.

File: frigate/prometheus_exporter.py
Comment (onedr0p) [reply frigate/prometheus_exporter.py:1]: @mitchross `stats.py` and `prometheus_metrics.py` seem fine.

File: frigate/prometheus_exporter.py
Comment (mweinelt) [reply frigate/prometheus_exporter.py:33]: Alright. I just felt this is the right place and time to iterate on the metrics. With that being said, here is my train of thought:

- Prometheus is primarily used for monitoring, so I assume monitoring the health of the process is more interesting than its actual pid value.
- The promql query for noticing a changed pid is more complex, than for an up metric
- What is the pid value, when there is no process?

Encoding the process health in an `up` metric would make much more sense to me and at the same time leave out a few sharp edges. Hence, why I feel this should be improved before merging it.

Comment (mitchross) [reply frigate/prometheus_exporter.py:33]: Could you provide the gist/pseudo code please? I can implement this but I don't really know precisely what you want here.

Comment (mweinelt) [reply frigate/prometheus_exporter.py:33]: It looks like pid values are `None` when the process does not exist.

https://github.com/blakeblackshear/frigate/blob/318240c14c36623d30724511640f3d3e16185e25/frigate/stats.py#L210-L218

Comment (mweinelt) [inline frigate/prometheus_exporter.py:33]:
```suggestion
        ffmpeg_up = GaugeMetricFamily(
            "frigate_ffmpeg_up", "Whether the ffmpeg process for a camera is up", labels=["camera"]
        )
```

File: frigate/prometheus_exporter.py
Comment (mweinelt) [inline frigate/prometheus_exporter.py:71]:
```suggestion
            add_metric(ffmpeg_up, k, d is not None, "ffmpeg_up")
```

File: frigate/prometheus_exporter.py
Comment (mweinelt) [inline frigate/prometheus_exporter.py:80]:
```suggestion
        yield ffmpeg_up
```

**Review by mweinelt — COMMENTED** [review]
In broad strokes, this is what should be required to convert the pid metric into an up metric.

The stats endpoint encodes `None` as json, the `CustomCollector` loads it as JSON, so matching the identity of `None` should be sufficient.

```python
>>> json.loads(json.dumps(None)) is None
True
```

File: frigate/prometheus_exporter.py
Comment (mweinelt) [reply frigate/prometheus_exporter.py:33]: If it restarts and doesn't stay down, that is great. But such an event is ultimately pretty uninteresting from a monitoring perspective where the service availability is of interest.

Comment (NickM-27) [reply frigate/prometheus_exporter.py:33]: > Edit: Oh, but it may be entirely unsuitable for noticing the process is actually down in that case.

Right, that is my point. Once frigate has started and the process is up, this will always show as the process being up even in the case where it continues to crash

Comment (NickM-27) [reply frigate/prometheus_exporter.py:33]: We are actually using the camera_fps for this case. The camera_fps will be set to 0 when the ffmpeg process has crashed.

I am not sure if the PID is useful for this particular application

Comment (mweinelt) [reply frigate/prometheus_exporter.py:33]: The process is a `multiprocessing.Process`, in which case `is_alive()` would be a better option.

Comment (mitchross) [reply frigate/prometheus_exporter.py:33]: When I get home today, i'll give this a whirl.

**Comment by mitchross** [comment]
Alright @mweinelt I think? I got your requests in based on your "pull request suggestions" ( I just chose submit/accept proposal )

If not, then perhaps we let this merge and you can submit a follow up pull request with tweaks/tunes to the prometheus class as you have more experience than I do here.

@NickM-27 I addressed refactoring of stats into a monitoring folder.

**Comment by NickM-27** [comment]
Can also make a PR pointed at your fork and then merge that there which will include the commits here.

To be clear, I added the pinned label because (unless Blake disagrees) I think this is something to hold off on merging until 0.12 is released. At this point 0.12 is feature complete and hopefully will advance to RC stage soon.

**Comment by mitchross** [comment]
> Can also make a PR pointed at your fork and then merge that there which will include the commits here.
> 
> To be clear, I added the pinned label because (unless Blake disagrees) I think this is something to hold off on merging until 0.12 is released. At this point 0.12 is feature complete and hopefully will advance to RC stage soon.

@mweinelt  what do you think?

**Comment by mweinelt** [comment]
Once I have set up, which may well take some months still, I can contribute things back.

@NickM-27 made a good point at https://github.com/blakeblackshear/frigate/pull/5594#discussion_r1120228118 about the process never being `None` after it was first set, which is why checking for `None` may not be a good idea after all.

**Comment by onedr0p** [comment]
@mweinelt I don't mean to sound rude and don't speak for the Frigate devs; however, while your comments are appreciated it would be nice if you left the review to the maintainers to check over so the author is not going back and forth between you supposed suggestions. If you want to help out that's cool, this can be merged in a working state (maybe not ideal to your standards) and if you want to layer on changes afterwards when you are free that's awesome.

**Comment by mweinelt** [comment]
I was trying to make a suggestion on how things could be improved given my experience with Prometheus. There was no intention to derail this pull request. It would have been fine to tell me, that this was out of scope.

**Comment by NickM-27** [comment]
I definitely don't have any experience with prometheus, and I believe Blake said he does not either, so it is important to get perspective and feedback from those that actually have used it so we can understand how this is working and what impacts different metrics / formats will have. These discussions are important to have.

There is also no rush to get the PR merged since it is a whole new feature (and restructuring) so I don't think it makes sense to be part of 0.12.

**Comment by onedr0p** [comment]
@bairhys (https://github.com/bairhys/prometheus-frigate-exporter) if you are around maybe you can also provide some insight here.

**Comment by onedr0p** [comment]
> I definitely don't have any experience with prometheus, and I believe Blake said he does not either, so it is important to get perspective and feedback from those that actually have used it so we can understand how this is working and what impacts different metrics / formats will have. 

Since you some some interest in accepting changes for adding prometheus metrics take a look at these pages:
- https://prometheus.io/docs/instrumenting/writing_exporters/
- https://github.com/prometheus/client_python#instrumenting

I can also do my best to help explaining this if you all have 30 minutes to an hour to spare in some sort of chat or voice or whatever. It really isn't that hard to understand and at a very high level you could wouldn't have a problem reviewing this.

**Comment by bairhys** [comment]
I have updated my exporter so that it labels the camera and detector process cpu and memory metrics. This makes them easier to identify compared to the other cpu and memory metrics.  I have removed the process id metrics for cameras and detectors since the cpu and memory metrics are labeled with pid values. Example metrics and grafana dashboard example can be found in the readme here https://github.com/bairhys/prometheus-frigate-exporter

Process CPU usage looks like this in Grafana

image

**Comment by mitchross** [comment]
> https://github.com/bairhys/prometheus-frigate-exporter

If you want to fork my PR and add your changes, id welcome it. Otherwise I can try to incorporate later.

**Comment by bairhys** [comment]
I added the changes here to your PR https://github.com/mitchross/frigate/pull/2 

It works fine in my exporter docker container but I haven't tested it with your PR yet

**Comment by mitchross** [comment]
> I added the changes here to your PR [mitchross#2](https://github.com/mitchross/frigate/pull/2)
> 
> It works fine in my exporter docker container but I haven't tested it with your PR yet

Approved!

**Comment by mitchross** [comment]
@blakeblackshear @NickM-27  resolved conflicts.

**Review by NickM-27 — CHANGES_REQUESTED** [review]
have a conflict from latest PR to fix and checks are not passing

**Comment by mitchross** [comment]
> have a conflict from latest PR to fix and checks are not passing

@NickM-27  should be good now.

File: docker-compose.yml
Comment (NickM-27) [inline docker-compose.yml:24]: I think more often than not nvidia isn't going to be used here, but we should have a comment explaining when this would need to be enabled

File: docker-compose.yml
Comment (NickM-27) [inline docker-compose.yml:34]:
```suggestion
      - /dev/bus/usb:/dev/bus/usb
```

File: requirements-wheels.txt
Comment (NickM-27) [inline requirements-wheels.txt:22]:
```suggestion
```

**Review by NickM-27 — CHANGES_REQUESTED** [review]

File: docker-compose.yml
Comment (mitchross) [reply docker-compose.yml:24]: my apologies! this file was NOT supposed to be committed.

Comment (mitchross) [reply docker-compose.yml]: @NickM-27 restored the original yaml

**Comment by NickM-27** [comment]
Need to fix conflicts

**Review by NickM-27 — APPROVED** [review]

File: frigate/monitoring/prometheus.py
Comment (d0ugal) [inline frigate/monitoring/prometheus.py:54]: I might be missing something - but why do a request to the Frigate API rather than importing what we need?

**Comment by d0ugal** [comment]
Is this change based on this code? https://github.com/bairhys/prometheus-frigate-exporter - if so it might be a problem that it has no license attached to it.

**Comment by onedr0p** [comment]
Seeing how the author of that repo directly helped here I don't think that's an issue.

**Comment by d0ugal** [comment]
> Seeing how the author of that repo directly helped here I don't think that's an issue.

Cool, hopefully not but figured I’d ask. IANAL but if I understand correctly permission needs to be explicitly granted otherwise it’s reserved.

File: frigate/monitoring/prometheus.py
Comment (blakeblackshear) [reply frigate/monitoring/prometheus.py:54]: I'm also curious about this. Any reason we need to use the http endpoint here?

Comment (mitchross) [reply frigate/monitoring/prometheus.py:54]: Its been a minute, I am trying to recollect.... I believe I couldn't find a way to pass the raw stats due to the life cycle of the Prometheus client. When the custom collector gets passed to the client start up we have this code 

image

Which will trigger

image

When the custom collector is registered, we then face

image

So now custom collector is called, thus "stats = json.loads(urlopen(self.stats_url).read())" is called


I remember fighting with this for days because I needed a way to bootstrap the Prometheus client with out it crashing. My mind is fuzzy here now, but there was for sure a issue with first time load and crashing. I recall reading all the tutorials and scanning GitHub to find examples and all of them did it this way. 

If have suggestions I'm all ears. If someone wants to open a PR against mine, Im welcome to that too.

Comment (bairhys) [reply frigate/monitoring/prometheus.py:54]: The use of the http stats is from my docker container. But this doesn't need to use the API to get the stats data. It could alternatively use a similar method to collect the stats as like how the API stats are generated (I haven't investigated this).

**Comment by bairhys** [comment]
I added a licence to https://github.com/bairhys/prometheus-frigate-exporter

File: frigate/monitoring/prometheus.py
Comment (mitchross) [reply frigate/monitoring/prometheus.py:54]: > The use of the http stats is from my docker container. But this doesn't need to use the API to get the stats data. It could alternatively use a similar method to collect the stats as like how the API stats are generated (I haven't investigated this).

Unable to get this to work. Only way I've gotten it to work is with the http route, for the same reasons I posted above. Im either missing something extremely easy here or the app is not architected to fit the Prometheus exporter library.

**Comment by fleaz** [comment]
Hey,
any plans on reviving this PR?
There was so much work put into this cool and helpful feature, it would be sad to let it rot.

**Comment by bairhys** [comment]
> Hey, any plans on reviving this PR? There was so much work put into this cool and helpful feature, it would be sad to let it rot.

You can use it right now via my docker container, see here: https://github.com/bairhys/prometheus-frigate-exporter

**Comment by mitchross** [comment]
redo-> https://github.com/blakeblackshear/frigate/pull/15514

**Pull request closed without merging**
